File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        
        vector<int> result;
        
        int rows = matrix.size();
        
        if (rows == 0) {
            return result;
        }
        
        int columns = matrix[0].size();
        
        // sr = start row, er = end row, sc = start column, ec = end column
        
        int sr = 0, sc = 0, er = rows-1, ec = columns-1;
        
        int count = 0;
        
        
        while (sr <= er && sc <= ec) {
            // print sr
            for(int i=sc; i<=ec; i++) {
                result.push_back(matrix.at(sr).at(i));
                count++;
            } 
            sr++;
            
            if (count==rows*columns) return result;
            
            for(int i=sr; i<=er; i++) {
                result.push_back(matrix.at(i).at(ec));
                count++;
            }
            ec--;
            
            if (count==rows*columns) return result;
            
            for(int i = ec; i>=sc; i--) {
                result.push_back(matrix.at(er).at(i));
                count++;
            }
            er--;
            
            if (count==rows*columns) return result;
            
            for(int i=er; i>=sr; i--)  {
                result.push_back(matrix.at(i).at(sc));
                count++;
            }
            sc++;
            
            if (count==rows*columns) return result;
        }
        
        return result;
        
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        sr, sc = 0, 0
        rows = len(matrix)
        cols = len(matrix[0])
        er = rows - 1
        ec = cols - 1
        
        result = []
        counter = 0
        
        while sr <= er or sc <= ec:
            for i in range(sc, ec+1):
                result.append(matrix[sr][i])
                counter += 1
            sr += 1
            if counter == rows*cols:
                return result
            
            for i in range(sr, er+1):
                result.append(matrix[i][ec])
                counter += 1
            ec -= 1
            if counter == rows*cols:
                return result
            
            for i in range(ec, sc-1, -1):
                result.append(matrix[er][i])
                counter += 1
            er -= 1
            if counter == rows*cols:
                return result
            
            for i in range(er, sr-1, -1):
                result.append(matrix[i][sc])
                counter += 1
            sc += 1
            if counter == rows*cols:
                return result
            
        
        return result